artemis-commons/src/main/java/org/apache/activemq/artemis/utils/collections/TypedProperties.java

```diff
@@ -372,6 +372,79 @@ private synchronized void forEachInternal(BiConsumer<SimpleString, PropertyValue
       }
    }
 
+   /**
+    * Performs a search among the valid key properties contained in {@code buffer}, starting from {@code from}
+    * assuming it to be a valid encoded {@link TypedProperties} content.
+    *
+    * @throws IllegalStateException if any not-valid property is found while searching the {@code key} property
+    */
+   public static boolean searchProperty(SimpleString key, ByteBuf buffer, int startIndex) {
```

Comment:
wouldn't be less costly to just lookup for the sequence of bytes?

You're still parsing the key at its exact position.

if just looking for the sequence , say _HDR_SQUEDULED_TIME anywhere in the buffer would be enough.

now if the search is more costly than the partial parsing, then it's fine.

At least you're not reading the string to verify the equal object.


if this was C, you would control to reuse the same integer every time, but in Java every time you do buffer.getInt(index) you will create a few bytes in memory.

Reply:
I have added a comment about this and the risk is to find a value that matches the property we are searching for. The other reason is that we know how many bytes to skip by using the decoded info and would help to reach quicker what we need without comparing every single byte (including values)

Reply:
if you could name this as hasProperty(SimpleString key).

since you use the parsed property if the buffer is not parsed already.

Reply:
I will change it for sure :+1

Reply:
>> "the risk is to find a value that matches the property we are searching for"

So what? if someone uses the key as the value, all it will incur is for us to do the older semantic, which is to parse the whole buffer. and we would incur in the old semantic.


Add a test with the value as the key, just to make sure.. although I doubt someone will use the property _HDR_SCHEDULED_TIME anywhere in the values :) and if they did it would still be fine.

Reply:
what I mean is.. if parsing the whole buffer is faster.. just go for it.. don't worry about values as keys.

Reply:
But is not faster according my tests...in one case we perform more comparisons...

Reply:
Re your comments

> You're still parsing the key at its exact position.

Nope: to compare `SimpleString` and `ByteBuf` is using another optimized method I've implemented some time ago that allow to compare them very quickly directly ie https://github.com/apache/activemq-artemis/blob/0293d8057441e6bfeb9e6533774862ffc86d74eb/artemis-commons/src/main/java/org/apache/activemq/artemis/utils/ByteUtil.java#L345 

> in Java every time you do buffer.getInt(index) you will create a few bytes in memory.

That's the signature of the Netty method: https://netty.io/4.0/api/io/netty/buffer/ByteBuf.html#getInt-int-

it would return a primitive `int`: it's not allocating anything. I've taken care to design that method to be garbage-free :+1: